source/extensions/filters/common/fault/fault_config.cc

```diff
@@ -64,6 +79,37 @@ absl::optional<Http::Code> FaultAbortConfig::HeaderAbortProvider::statusCode(
   return ret;
 }
 
+absl::optional<Grpc::Status::GrpcStatus> FaultAbortConfig::HeaderAbortProvider::grpcStatusCode(
+    const Http::RequestHeaderMap* request_headers) const {
+  if (request_headers == nullptr) {
+    return absl::nullopt;
+  }
+
+  auto header = request_headers->get(Filters::Common::Fault::HeaderNames::get().AbortGrpcRequest);
+  if (header == nullptr) {
+    return absl::nullopt;
+  }
+
+  uint64_t code;
+  if (!absl::SimpleAtoi(header->value().getStringView(), &code)) {
+    return absl::nullopt;
+  }
+
+  return static_cast<Grpc::Status::GrpcStatus>(code);
+}
+
+envoy::type::v3::FractionalPercent FaultAbortConfig::HeaderAbortProvider::percentage(
```

Comment:
I'm not sure whether with this implementation abort faults always work the way we want them to work. Let me give you an example that I was able to come up with and you may let me know whether my concerns are valid (in this case we should probably fix the issue) or whether I'm missing something here.

Example: 

Let's say that I send a request with the following HTTP headers: 
- `x-envoy-fault-abort-grpc-request-percentage`: `20`
- `x-envoy-fault-abort-request`: `500`
and a proto config telling Envoy to look for faults in HTTP headers.

Behavior:
There is 20% change that we apply 500 HTTP status code abort fault.

Expected behavior:
Per our documentation, in this case we should ingest HTTP status code abort fault in % of requests that's controlled by percentage field from our config file.

Reply:
I'm not sure whether you will need to do this but you may need to create separate fault providers for HTTP and GRPC aborts.

Reply:
In the current implementation, `x-envoy-fault-abort-grpc-request-percentage` value will be used only when `x-envoy-fault-abort-grpc-request` header is passed in. For the above example, we will end up using the percentage value from config file as mentioned in the documentation.

Here is the control logic for it, let me know if the comment and code around it is not clear, I can reword it. I'll add a unit test to capture this scenario.

`// if the abort fault is for grpc status, then use grpc fault request percentage.`

https://github.com/envoyproxy/envoy/blob/4576e1941980a4c13a2d800b03717f3154876791/source/extensions/filters/common/fault/fault_config.cc#L103

Reply:
I see 👍 Thanks for pointing this out - your are right, it works just fine as it is now. The only thing that I could be a little bit picky about here is the fact that we read the value `x-envoy-fault-abort-grpc-request` HTTP header twice and apparently (see [this convo](https://github.com/envoyproxy/envoy/pull/10724#discussion_r406981261) for more details) reading HTTP headers is not cheap performance wise. 

Maybe we could read `x-envoy-fault-abort-request` and `x-envoy-fault-grpc-request` headers first and pass it to `percentage` method for it to determine which percentage should be used? Better yet, if we read `x-envoy-fault-abort-request` first and it's present we don't even need to check the value of `x-envoy-fault-abort-grpc-request`

Reply:
In this particular case it's not a blocker tho - just something to think about (maybe we can address it if it's easily fixable)

Reply:
Thanks @Augustyniak, I couldn’t find a clean and easy way to avoid this check, so leaving it as is, but I’ve changed the precedence of http_status and grpc_status per recommendation.